### PR TITLE
user-detach and graceful user creation  were implemented

### PR DESCRIPTION
**Pull request by tengu-alt**
user-detach on cluster deletion and graceful user creation on cluster create were implemented

### CODE REVIEW

File: controllers/clusters/kafka_controller.go [hunk 1]
```diff
@@ -710,6 +779,70 @@ func (r *KafkaReconciler) newWatchStatusJob(kafka *v1beta1.Kafka) scheduler.Job
...
+
+		err := r.Get(ctx, types.NamespacedName{
+			Namespace: kafka.Namespace,
+			Name:      kafka.Name,
+		}, kafka)
+
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				l.Info("Kafka cluster resource is not found. Stopping user creation job...")
+				go r.Scheduler.RemoveJob(kafka.GetJobID(scheduler.UserCreator))
```
Comment (worryg0d) [inline controllers/clusters/kafka_controller.go:805]: I suggest you move this to the `watchClusterStatus` job.

File: controllers/clusters/kafka_controller.go [hunk 2]
```diff
@@ -382,6 +393,54 @@ func (r *KafkaReconciler) handleDeleteUser(
...
+			return err
+		}
+
+		l.Error(err, "Cannot get Kafka user", "user", u.Spec)
+		r.EventRecorder.Eventf(kafka, models.Warning, models.DeletionFailed,
+			"Cannot get Kafka user. user reference: %v", userRef)
+
+		return err
+	}
+	if _, exist := u.Status.ClustersEvents[kafka.Status.ID]; !exist {
```
Comment (worryg0d) [inline controllers/clusters/kafka_controller.go:423]: Add here new line and check similar cases and fix them if they exist.

**Review by worryg0d — APPROVED** [review]

**Review by worryg0d — COMMENTED** [review]
Oops, I've misclicked the `comment` button.

File: controllers/clusters/kafka_controller.go [hunk 3]
```diff
@@ -710,6 +779,70 @@ func (r *KafkaReconciler) newWatchStatusJob(kafka *v1beta1.Kafka) scheduler.Job
...
+				)
+				r.EventRecorder.Eventf(kafka, models.Warning, models.CreationFailed,
+					"Failed to create a user for the cluster. Reason: %v", err,
+				)
+
+				return err
+			}
+
+		}
+		go r.Scheduler.RemoveJob(kafka.GetJobID(scheduler.UserCreator))
```
Comment (ribaraka) [inline controllers/clusters/kafka_controller.go:835]: please move it after _finished job_ log

File: controllers/kafkamanagement/kafkauser_controller.go [hunk 4]
```diff
@@ -297,6 +297,61 @@ func (r *KafkaUserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
...
+			}
+
+			l.Info("Kafka user has been deleted",
+				"initial permissions", user.Spec.InitialPermissions,
+				"kafka user options", user.Spec.Options,
+			)
+
+			r.EventRecorder.Eventf(
+				user, models.Normal, models.Deleted,
+				"Resource is deleted",
```
Comment (ribaraka) [inline controllers/kafkamanagement/kafkauser_controller.go:326]: Resource is deleted -> User is detached from cluster

File: controllers/kafkamanagement/kafkauser_controller.go [hunk 5]
```diff
@@ -297,6 +297,61 @@ func (r *KafkaUserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
...
+				r.EventRecorder.Eventf(
+					user, models.Warning, models.PatchFailed,
+					"Resource status patch is failed. Reason: %v",
+					err,
+				)
+
+				return models.ReconcileRequeue, nil
+			}
+
+			l.Info("Kafka user has been deleted",
```
Comment (ribaraka) [inline controllers/kafkamanagement/kafkauser_controller.go:319]: deleted -> detached/removed

**Review by ribaraka — CHANGES_REQUESTED** [review]

File: controllers/clusters/kafka_controller.go (hunk 2 above)
Comment (tengu-alt) [reply controllers/clusters/kafka_controller.go:423]: done

File: controllers/clusters/kafka_controller.go [hunk 6]
```diff
@@ -559,6 +619,14 @@ func (r *KafkaReconciler) handleDeleteCluster(ctx context.Context, kafka *v1beta
 		return models.ReconcileRequeue
 	}
 
+	for _, ref := range kafka.Spec.UserRefs {
+		err = r.detachUser(ctx, kafka, l, ref)
+		if err != nil {
+
+			return models.ReconcileRequeue
+		}
+	}
```
[the remark this answers is not in the document]
Comment (tengu-alt) [reply controllers/clusters/kafka_controller.go:628]: done

File: controllers/clusters/kafka_controller.go (hunk 3 above)
Comment (tengu-alt) [reply controllers/clusters/kafka_controller.go:835]: done

File: controllers/kafkamanagement/kafkauser_controller.go (hunk 4 above)
Comment (tengu-alt) [reply controllers/kafkamanagement/kafkauser_controller.go:326]: done

File: controllers/kafkamanagement/kafkauser_controller.go (hunk 5 above)
Comment (tengu-alt) [reply controllers/kafkamanagement/kafkauser_controller.go:319]: done

**Review by testisnullus — APPROVED** [review]

**Review by ribaraka — APPROVED** [review]

**Review by taaraora — APPROVED** [review]

**Pull request merged**